calico/calcollections.py

```diff
@@ -75,3 +75,89 @@ def resulting_size(self):
     @property
     def empty(self):
         return not (self.added_entries or self.removed_entries)
+
+
+class OneToManyIndex(object):
+    """
+    Represents a mapping from key to a set of values.
+    """
+
+    def __init__(self, set_cls=set):
+        self._set_cls = set_cls
+        self._index = {}
+
+    def add(self, key, value):
+        """Add a mapping to the index.
+
+        Idempotent: does nothing if the mapping is already present.
+        """
+        index_entry = self._index.setdefault(key, value)
+        if index_entry != value:
+            # Failed to insert the new value as the single entry, examine
+            # what we got.
+            if isinstance(index_entry, self._set_cls):
+                # Already have multiple values for that entry, add the new one.
+                index_entry.add(value)
+            else:
+                # There was an entry but it wasn't the one we tried to add,
+                # promote the entry to a set.
+                index_entry = self._set_cls([index_entry, value])
+                self._index[key] = index_entry
+
+    def discard(self, key, value):
+        """Discards a mapping from the index.
+
+        Idempotent: does nothing if the mapping is already gone.
+        """
+        if key in self._index:
+            index_entry = self._index[key]
+            if isinstance(index_entry, self._set_cls):
+                index_entry.discard(value)
+                if len(index_entry) == 1:
+                    index_entry = index_entry.pop()
+                    self._index[key] = index_entry
+            elif index_entry == value:
+                del self._index[key]
+
+    def __contains__(self, item):
+        return item in self._index
+
+    def contains(self, key, value):
+        """
+        :return: True if the given key/value mapping is present.
+        """
+        index_entry = self._index.get(key)
+        if isinstance(index_entry, self._set_cls):
+            return value in index_entry
+        else:
+            return value == index_entry
+
+    def iter_values(self, key):
+        """
+        :return: an iterator over the values for the given key.  WARNING:
+                 care should be taken not o modify the values associated with
+                 that key while iterating.
+        """
+        if key in self._index:
+            index_entry = self._index[key]
+            if isinstance(index_entry, self._set_cls):
+                return iter(index_entry)
+            else:
+                return iter([index_entry])
+        return iter([])
+
+    def num_items(self, key):
+        """
+        :return: The number of items associated with the given key.  Returns 0
+                 if the key is not in the mapping.
```

Comment:
I think the optimization is worthwhile here.  All this code is on the critical path so avoiding iterating over a lot of items in the worst case is beneficial.